Scrolling/pagination controls

aslagle opened the pull request:
This has the first part of #6 and also fixes #5.

It adds a settings button to the controls that brings up links to toggle between scrolling and pagination.

aslagle committed:
Added settings and scrolling view option.

aslagle committed:
Changed interfaces to use constructors or factory methods instead of start, and got rid of WebpubViewer class. (Fixes #5)

aslagle committed:
Save selected view.

aslagle committed:
Added a CSS class to the link for the current selected view.

aslagle committed:
Moved scrolling out of iframe so saving scroll position can work on ios.

aslagle commented on examples/alice/webpub-viewer.js:
```diff
@@ -0,0 +1 @@
+../../dist/webpub-viewer.js
```
This was to make it easier to test the service worker.

winniequinn commented on examples/alice/index.html:
```diff
@@ -53,15 +57,19 @@
-      viewer.start(element);
+      var manifestUrl = new URL("manifest.json", window.location.href).href;
+
+      var store = new LocalStorageStore.default(manifestUrl);
```
It might be better if `manifestUrl` were actually of type `URL`. It would ensure that `LocalStorageCache`, `ServiceWorkerCacher`, and so on are at least initialized with well-formed URLs. They could, of course, internally call the `href` method on the URL if they only need it as a string.

winniequinn commented on examples/alice/index.html:
```diff
@@ -53,15 +57,19 @@
+        var paginator = new ColumnsPaginatedBookView.default();
+        var scroller = new ScrollingBookView.default();
+        var annotator = new LocalAnnotator.default(store);
+        BookSettings.default.create([paginator, scroller], store).then(function (settings) {
```
I really like how initialization was changed to use promises!

aslagle replied:
yay!

winniequinn commented on src/BookSettings.ts:
```diff
@@ -0,0 +1,120 @@
+    private viewLinks: { [key: string]: HTMLAnchorElement };
+    private selectedView: BookView;
+    private viewChangeCallback: () => void;
+    private SELECTED_VIEW_KEY = "settings-selected-view";
```
You can declare `SELECTED_VIEW_KEY` with the `private static readonly` qualifiers and change all references from `this.SELECTED_VIEW_KEY` to `BookSettings. SELECTED_VIEW_KEY`. This should actually give you a proper constant!

winniequinn commented on src/BookSettings.ts:
```diff
@@ -0,0 +1,120 @@
+`;
+
+export default class BookSettings extends HTMLView {
+    private store: Store;
```
Many of these properties are only set in the constructor. As such, you can declare them `private readonly` which will make this explicit. (The `readonly` qualifier restricts mutation to occurring within the constructor of the class only. It is essentially `const` for instance variables.)

winniequinn commented on src/BookSettings.ts:
```diff
@@ -0,0 +1,120 @@
+    }
+
+    public renderControls(element: HTMLElement): void {
+        let sections = [];
```
Also picky, but you can declare `sections` as `const` here: You're mutating the array `sections` is initialized to, not the variable `sections itself.

winniequinn commented on src/BookSettings.ts:
```diff
@@ -0,0 +1,120 @@
+            let options = [];
+            for (const bookView of this.bookViews) {
+                options.push(optionTemplate(bookView.name, bookView.label));
+            }
```
There's a more pleasant and functional way to write this, for what it's worth:

```typescript
const options = this.bookViews.map(bookView =>
    optionTemplate(bookView.name, bookView.label)
);
```

winniequinn commented on src/BookSettings.ts:
```diff
@@ -0,0 +1,120 @@
+                    this.selectedView = view;
+                    this.updateViewLinks();
+                    this.storeSelectedView(view);
+                    if (this.viewChangeCallback()) {
```
I think you meant `if (this.viewChangeCallback)` here: I don't think you want to call it to see if you have it!

winniequinn commented on src/BookView.ts:
```diff
@@ -0,0 +1,11 @@
+    label: string;
+    setBookElement(element: Element): void;
+    setTopMargin(number: number): void;
+    start(position: number): void;
```
It would be good to document what `start` and `stop` are meant to do here.

aslagle replied:
I tried.... it's hard to explain.

winniequinn commented on src/ColumnsPaginatedBookView.ts:
```diff
@@ -1,13 +1,24 @@
 
-    public start(iframe: HTMLIFrameElement, position: number): Promise<void> {
+    public name = "columns-paginated-view"
+    public label = "Paginated View"
```
These should be `readonly`, I think!

winniequinn commented on src/HTMLView.ts:
```diff
@@ -0,0 +1,14 @@
+        return parentElement.querySelector(selector);
+    }
+
+    protected findRequiredElement(parentElement: Element, selector: string): Element {
```
It would be good to document that this method may throw an exception.

winniequinn commented on src/HTMLView.ts:
```diff
@@ -0,0 +1,14 @@
+export default class HTMLView {
```
I am not sure `HTMLView` should be a class. It seems like it is just a couple of utility methods.

Perhaps it would be better to change `HTMLView` to be a module with two exported functions and change `BookSettings` and `IFrameNavigator` to simply use the functions as needed? Nothing seems to depend on anything being a subclass of `HTMLView`.

If you do want it to remain a class for some reason, I would consider declaring its two methods `static`.

winniequinn reviewed (changes requested):
I noticed a weird visual bug in desktop Safari where toggling the pagination mode causes the outermost bullet to disappear. Everything else appears to be okay though!

aslagle committed:
Use URL type for manifest URL.

aslagle committed:
Use readonly for things that are only set in the constructor.

aslagle committed:
Arrays can be declared as const.

aslagle committed:
Use map to create options html.

aslagle committed:
Fixed extra call of viewChangeCallback.

aslagle committed:
Document BookView start and stop.

aslagle committed:
Made HTMLUtilities a module instead of a class.

aslagle commented:
@winniequinn I addressed most of your comments!

I'm not too worried about the bullet disappearing since the styling will be changing significantly.

I opened a separate issue (#10) for the other bug you found and I may come back to it after I implement margins.
